specs/logging_implementation.adoc

```diff
@@ -0,0 +1,262 @@
+* Ideally logs messages, operation updates and alerts are messages which
+  are emitted from an application.
+
+== Use Cases
+
+Logs and alerts need to be treated as messages.
+
+== Proposed change
+* Logs, alerts and operations need to be treated as messages emitted by Tendrl modules.
+* A schema needs to be defined for emitted messages.
+* Message priorities, types etc. need to be defined at the time of emission.
+* The integration/bridge applications must pass the message as per schema.
+* For messages that serve as notifications to the administrator (or alerts),
+  exact error message must be hardcoded by the developers. No backtraces or
+  system generated error messages should be sent as notifications to the
+  administrator.
+* Notifications sent to the administrator should provide sufficient context to
+  enable debugging of the problem via investigation into the log files.
+* Notifications and alerts need to be delivered to the node agent alert sockets.
+* Common module needs to implement a logging library that hides all the implementation
+  details and configuration options from any of the components using the library.
+  Components should be able to blindly emit messages to the Message Handler module
+  in Tendrl and it should be able to deliver them to either syslog, log files,
+  alert sockets etc. depending upon the metadata and configuration.
+* Syslog must be the default logging configuration. Logging to individual
+  configuration files should be thoroughly discouraged.
+
+== Alternatives
+
+* None
+
+== Data model impact
+
+* Store the operation details in etcd:
+```
+1. Operation messages with inorder_id
+    key: {req_id}/{in_order_id}
+
+    value: {
+       message_id : "uuid",
+       priority: "", (no priority)
+       timestamp: time_now,
+       publisher: "node_agent/ceph_integration......" (particular module name)
+       payload: {
+          "job_id": "",
+          "node_id": "",
+          "cluster_id": "",
+          "flow_id": "",
+          "parent_id",
+          "message": ""
+       }
+    }
+
+2. Messages (except operation, info and debug) stored in etcd
+   keys:
+    In global level : /messages/{message_id}
+    In node level   : /nodes/{node_id}/messages/{message_id}
+    In cluster level: /clusters/{cluster_id}/messages/{message_id}
+
+  value:{
+      message_id : "uuid",
+      priority: "", (no priority)
+      timestamp: time_now,
+      publisher: "node_agent/ceph_integration......" (particular module name)
+      payload: {
+         "node_id": "",
+         "cluster_id": "",
+         "message": """
+      }
+   }
+
+    (operational messages stored in operational directory only)
+```
+
+== Impacted modules
+
+=== Tendrl/common impact
+* Create a logger module in common to handle the message handler object.
+* Logging module will takecare of log the messages in log file.
+* Create a new function called emit in logger module to process message object.
```

Comment:
emit = generate, handle = process. What precisely do you mean for the emit method to do?